### ISSUE TITLE
Argument missing for parameter

### ISSUE THREAD

**Issue by lexicalunit** [django/asgiref#394]
There's a bug with the type hinting in version 3.7.0+

```python
from asgiref.sync import sync_to_async


class Whatever:
    @sync_to_async()
    def something(self, i: int) -> int:
        return i + 1


what = Whatever()


async def main() -> None:
    await what.something(i=1)  # error is here
```

Produces the following error in `pyright`:

> Argument missing for parameter "self" [reportGeneralTypeIssues]

And with `mypy`:

> error: Too few arguments  [call-arg]

This code did not produce any type hinting issues in version 3.6.0 and prior.

- pyright 1.1.309
- mypy 1.30
- python 3.10.10

**Comment by bellini666** [comment]
There are some typing changed that I did in [this PR](https://github.com/django/asgiref/pull/390/files)

But it is strange because I can't reproduce your issue at all. I tested your code with both python 3.10 and 3.11 and mypy/pyright in their latest versions, and both worked fine without issues. Even by doing a `reveal_type` at the returned value of the await call it says that it is an `int`

Can you check if the issue happens on 3.7.1?
Also, do you have custom settings for pyright/mypy?

**Comment by lexicalunit** [comment]
Just to make sure I'm not using any weird configuration or anything like that, I created a script that generates a new virtualenv and installs only pyright and asgiref, generates the some example code, and runs pyright against it:

```shell
#!/bin/bash -ue

mkdir test
cd test
python3.10 -m venv env
source env/bin/activate
pip install "pyright==1.1.309"
pip install "asgiref==3.7.1"
cat >test.py <<EOF
from asgiref.sync import sync_to_async

class Whatever:
    @sync_to_async()
    def something(self) -> None:
        return None

what = Whatever()

async def main() -> None:
    await what.something()  # error is here
EOF
source env/bin/activate
echo "================"
echo -n "which pyright:"; which pyright
echo -n "which python:"; which python
echo -n "pyright --version:"; pyright --version
echo -n "python --version:"; python --version
pyright --verbose test.py
```

Here's all output I get from executing the script:

```log
➜ ./test.sh
Collecting pyright==1.1.309
  Using cached pyright-1.1.309-py3-none-any.whl (17 kB)
Collecting nodeenv>=1.6.0
  Using cached nodeenv-1.8.0-py2.py3-none-any.whl (22 kB)
Requirement already satisfied: setuptools in ./env/lib/python3.10/site-packages (from nodeenv>=1.6.0->pyright==1.1.309) (65.4.1)
Installing collected packages: nodeenv, pyright
Successfully installed nodeenv-1.8.0 pyright-1.1.309

[notice] A new release of pip available: 22.2.2 -> 23.1.2
[notice] To update, run: pip install --upgrade pip
Collecting asgiref==3.7.1
  Using cached asgiref-3.7.1-py3-none-any.whl (24 kB)
Collecting typing-extensions>=4
  Using cached typing_extensions-4.6.2-py3-none-any.whl (31 kB)
Installing collected packages: typing-extensions, asgiref
Successfully installed asgiref-3.7.1 typing-extensions-4.6.2

[notice] A new release of pip available: 22.2.2 -> 23.1.2
[notice] To update, run: pip install --upgrade pip
================
which pyright:/Users/lexicalunit/test/env/bin/pyright
which python:/Users/lexicalunit/test/env/bin/python
pyright --version:WARNING: there is a new pyright version available (v1.1.309 -> v1.1.310).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

pyright 1.1.309
python --version:Python 3.10.8
WARNING: there is a new pyright version available (v1.1.309 -> v1.1.310).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

Search paths for /Users/lexicalunit/test
  /Users/lexicalunit/.cache/pyright-python/1.1.309/node_modules/pyright/dist/typeshed-fallback/stdlib
  /Users/lexicalunit/test
  /Users/lexicalunit/test/typings
  /Users/lexicalunit/.cache/pyright-python/1.1.309/node_modules/pyright/dist/typeshed-fallback/stubs/...
  /usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10
  /usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/lib-dynload
  /Users/lexicalunit/test/env/lib/python3.10/site-packages
Found 1 source file
pyright 1.1.309
/Users/lexicalunit/test/test.py
  /Users/lexicalunit/test/test.py:11:11 - error: Argument missing for parameter "self" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
Completed in 0.615sec
```

**Comment by bellini666** [comment]
@lexicalunit ok, I think this is an issue with pyright on version 1.1.309. I could replicate your issue with that version

I'm using pyright installed from `npm` directly on 1.1.310 and in that version it has no errors (probably something that was fixed on 1.1.310). If I downgrade it by `npm i -g pyright@1.1.309` than the same issue happens, but stops happening after doing `npm i -g pyright@1.1.310`

Regarding mypy, I still get no errors with it as well

**Comment by lexicalunit** [comment]
Sorry I wasn't explicit. @bellini666, my above post was using pyright 1.1.310 and I saw the error.

I was able to get the error to go away by changing this line:

```python
    @sync_to_async
```

to this:

```python
    @sync_to_async()
```

Here's the full script which shows a passing test:

```shell
#!/bin/bash -ue

mkdir test
cd test
python3.10 -m venv env
source env/bin/activate
pip install "pyright==1.1.310"
pip install "asgiref==3.7.1"
cat >test.py <<EOF
from asgiref.sync import sync_to_async

class Whatever:
    def __init__(self):
        pass

    @sync_to_async()
    def something(self) -> None:
        return None

what = Whatever()

async def main() -> None:
    await what.something()  # error is here
EOF
source env/bin/activate
echo "================"
echo -n "which pyright:"; which pyright
echo -n "which python:"; which python
echo -n "pyright --version:"; pyright --version
echo -n "python --version:"; python --version
pyright --verbose test.py
```
